### ISSUE TITLE
Wrong operation in stft's center attribute

### ISSUE THREAD

**Issue by lbjcom** [pytorch/pytorch#72328]
### 🐛 Describe the bug

According to the pytorch's document, if I set `center` to `True` the first frame should be centered in the beginning of stft:
https://github.com/pytorch/pytorch/blob/2367face24afb159f73ebf40dc6f23e46132b770/torch/functional.py#L620

And the following code shows that the stft pads by `n_fft//2`
https://github.com/pytorch/pytorch/blob/2367face24afb159f73ebf40dc6f23e46132b770/torch/functional.py#L692

This works fine in most cases but if the `win_length` and `n_fft` are different padding `n_fft//2` does not gurantee the first frame be centered. I think `win_length//2` is the right choice for the proper operation.

### Versions

Collecting environment information...
PyTorch version: 1.10.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Sep 28 2021, 16:10:42)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-693.5.2.el7.x86_64-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Tesla P40
GPU 1: Tesla P40
GPU 2: Tesla P40
GPU 3: Tesla P40
GPU 4: Tesla P40
GPU 5: Tesla P40
GPU 6: Tesla P40
GPU 7: Tesla P40

Nvidia driver version: 455.32.00
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.4
[pip3] pytorch-lightning==1.4.9
[pip3] torch==1.10.1+cu113
[pip3] torchaudio==0.10.1+cu113
[pip3] torchmetrics==0.6.1
[pip3] torchvision==0.11.2+cu113
[conda] Could not collect

cc @brianjo @mruberry @peterbell10 @ezyang @gchanan @zou3519 @anjali411 @dylanbespalko @Lezcano @nikitaved

**Comment by peterbell10** [comment]
When `win_length < n_fft`, the window is padded at both ends so the window is always centred in `n_fft`. I think it follows that padding by `n_fft//2` is right for all window sizes. Relevant code:
https://github.com/pytorch/pytorch/blob/8f2e7bba831ef9b7013d673870b162056cb51fcf/aten/src/ATen/native/SpectralOps.cpp#L852-L860

Let me know if you disagree.

**Comment by lbjcom** [comment]
@peterbell10 Thanks for the reply. Your code shows that the "**window** is centered". What I want to show you is the following comment:
https://github.com/pytorch/pytorch/blob/2367face24afb159f73ebf40dc6f23e46132b770/torch/functional.py#L620-L623

I think "window is centered" and the "frame is centered" have different meaning.

**Comment by lbjcom** [comment]
@peterbell10 
I didn't catch your point at first but I've just understood it. Both padding in https://github.com/pytorch/pytorch/blob/2367face24afb159f73ebf40dc6f23e46132b770/torch/functional.py#L692 and https://github.com/pytorch/pytorch/blob/8f2e7bba831ef9b7013d673870b162056cb51fcf/aten/src/ATen/native/SpectralOps.cpp#L852-L860 gurantees 0th sample be centered :)

Thanks!

**Comment by lbjcom** [comment]
Sorry but here's another question 😭 
What if `center` is False? Then input is not padded so window may miss first `n_fft-win_length` samples of input.

**Comment by lbjcom** [comment]
@peterbell10 Here's an example for you:

```python
def test_win_length_padding1():
    input1 = torch.cat((torch.ones(300), torch.zeros(618)))
    input2 = torch.cat((torch.ones(400), torch.zeros(518)))

    out1 = torch.stft(input1, n_fft=512, win_length=300, hop_length=1024, center=False).squeeze_()
    out2 = torch.stft(input2, n_fft=512, win_length=300, hop_length=1024, center=False).squeeze_()

    out1 = out1[:,0]*out1[:,0]+out1[:,1]*out1[:,1]
    out2 = out2[:,0]*out2[:,0]+out2[:,1]*out2[:,1]

    assert torch.isclose(out1, out2).all()


def test_win_length_padding2():
    input1 = torch.cat((torch.zeros(106), torch.ones(300), torch.zeros(618)))
    input2 = torch.cat((torch.ones(406), torch.zeros(618)))

    out1 = torch.stft(input1, n_fft=512, win_length=300, hop_length=1024, center=False).squeeze_()
    out2 = torch.stft(input2, n_fft=512, win_length=300, hop_length=1024, center=False).squeeze_()

    out1 = out1[:,0]*out1[:,0]+out1[:,1]*out1[:,1]
    out2 = out2[:,0]*out2[:,0]+out2[:,1]*out2[:,1]

    assert torch.isclose(out1, out2).all()
```

`test_win_length_padding1()` has to be successful if `center=False` works properly and 0th frame start from the first sample. However `test_win_length_padding1()` fails and `test_win_length_padding2()` is successful which is padded by 106 samples (`(n_fft - win_length)/2`), which means that `torch.stft` skips first `(n_fft - win_length)/2` if `center=False`.

**Comment by peterbell10** [comment]
If the window is zero padded and the input signal isn't, then you're guaranteed to miss some data either at the start or the end of the input. Center padding means you lose half as much at the start and the other half at the end. Right padding would mean you lose it all at the end.

**Comment by peterbell10** [comment]
Closing because I think the documentation is correct here and even includes that the window is center-padded:
>window will be padded on both sides to length `n_fft` before being applied.

**Comment by lbjcom** [comment]
> If the window is zero padded and the input signal isn't, then you're guaranteed to miss some data either at the start or the end of the input.

Then I have to use `center=True` for keeping all start and end data safe. Thanks!

And, I think adding [the above comment](https://github.com/pytorch/pytorch/issues/72328#issuecomment-1031520936) to the document would be helpful for users. In my case I didn't expect missing input data in any case.
